module/state_synchronization/indexer/persister.go

```diff
@@ -0,0 +1,207 @@
+	inMemoryResults        *unsynchronized.LightTransactionResults
+	inMemoryTxResultErrMsg *unsynchronized.TransactionResultErrorMessages
+	registers              storage.RegisterIndex
+	events                 storage.Events
+	collections            storage.Collections
+	transactions           storage.Transactions
+	results                storage.LightTransactionResults
+	txResultErrMsg         storage.TransactionResultErrorMessages
+	executionResult        *flow.ExecutionResult
+	header                 *flow.Header
+
+	LastPersistedSealedExecutionResult *flow.ExecutionResult
+}
+
+// NewPersister creates a new persister for transferring data from in-memory storage to permanent storage.
+func NewPersister(
+	log zerolog.Logger,
+	inMemoryRegisters *unsynchronized.Registers,
+	inMemoryEvents *unsynchronized.Events,
+	inMemoryCollections *unsynchronized.Collections,
+	inMemoryTransactions *unsynchronized.Transactions,
+	inMemoryResults *unsynchronized.LightTransactionResults,
+	inMemoryTxResultErrMsg *unsynchronized.TransactionResultErrorMessages,
+	registers storage.RegisterIndex,
+	events storage.Events,
+	collections storage.Collections,
+	transactions storage.Transactions,
+	results storage.LightTransactionResults,
+	txResultErrMsg storage.TransactionResultErrorMessages,
+	executionResult *flow.ExecutionResult,
+	header *flow.Header,
+) *Persister {
+	persister := &Persister{
+		log:                    log.With().Str("component", "persister").Logger(),
+		inMemoryRegisters:      inMemoryRegisters,
+		inMemoryEvents:         inMemoryEvents,
+		inMemoryCollections:    inMemoryCollections,
+		inMemoryTransactions:   inMemoryTransactions,
+		inMemoryResults:        inMemoryResults,
+		inMemoryTxResultErrMsg: inMemoryTxResultErrMsg,
+		registers:              registers,
+		events:                 events,
+		collections:            collections,
+		transactions:           transactions,
+		results:                results,
+		txResultErrMsg:         txResultErrMsg,
+		executionResult:        executionResult,
+		header:                 header,
+	}
+
+	persister.log.Info().
+		Uint64("height", header.Height).
+		Msg("persister initialized")
+
+	return persister
+}
+
+// AddToBatch adds data from in-memory storage to the provided batch.
+// It processes events, transaction results, registers, collections, and transactions
+// from their respective in-memory storages to permanent storage.
+// Note: This method doesn't commit the batch - that should be done by the caller.
+func (p *Persister) AddToBatch(batch storage.Batch) error {
+	log := p.log.With().
+		Hex("block_id", logging.ID(p.executionResult.BlockID)).
+		Uint64("height", p.header.Height).
+		Logger()
+
+	log.Debug().Msg("adding execution data to batch")
+
+	start := time.Now()
+
+	if err := p.addEventsToBatch(batch); err != nil {
+		return err
+	}
+
+	if err := p.addResultsToBatch(batch); err != nil {
+		return err
+	}
+
+	if err := p.persistRegisters(); err != nil {
```

Comment:
this is run on a separate database and is not included in the batch. run it first so we can skip the batch store if it fails.